### ISSUE TITLE
Missing Method for Gradients involving * (?)

### ISSUE THREAD

**Issue by arnavs** [JuliaDiffEq/ModelingToolkit.jl#208]
See below. 

```
julia> typeof(∇F[4])
Operation

julia> typeof(θᵉ)
Array{Operation,1}

julia> ∇F[4]
((b * a * exp(z_p) * k_p ^ (a - 1)) / c_p) / c_p

julia> θᵉ
2-element Array{Operation,1}:
 r
 b

julia> ModelingToolkit.gradient(∇F[4], θᵉ)
ERROR: MethodError: no method matching derivative(::typeof(*), ::NTuple{4,Operation}, ::Val{1})
Closest candidates are:
  derivative(::typeof(*), ::Tuple{Any,Any}, ::Val{1}) at /Users/arnavsood/.julia/packages/ModelingToolkit/tNeiU/src/differentials.jl:104
  derivative(::Operation, ::Any) at /Users/arnavsood/.julia/packages/ModelingToolkit/tNeiU/src/differentials.jl:97
  derivative(::typeof(asind), ::Tuple{Any}, ::Val{1}) at /Users/arnavsood/.julia/packages/ModelingToolkit/tNeiU/src/differentials.jl:104
```

**Comment by wupeifan** [comment]
Is it because NTuple is not Tuple? The first closet candidate should be the one?

**Comment by arnavs** [comment]
@wupeifan This is actually a weird one to get a MWE for. It seems to work fine in global scope... the other thing is that `@parameter` vs `@variable` makes things semantically different (i.e., if you do `@parameters r b`, then all the derivatives go to 0... but if you do it `@variables`, then it does the correct symbolic derivative.)

But I suspect you're right, about the actual method missing here.

**Comment by arnavs** [comment]
For example, here it is in the REPL, which is fine... but when the expression is the output of a `ModelingToolkit.jacobian()` call inside of a loop... then things break. 

```
julia> using ModelingToolkit

julia> @variables c c_p z z_p k k_p
(c, c_p, z, z_p, k, k_p)

julia> @parameters r b a sigma_2
(r, b, a, sigma_2)

julia> F = ((b * a * exp(z_p) * k_p ^ (a - 1)) / c_p) / c_p

((((b * a) * exp(z_p)) * k_p ^ (a - 1)) / c_p) / c_p

julia> ModelingToolkit.gradient(F, [r, b])
2-element Array{Expression,1}:
                        ModelingToolkit.Constant(0)
 exp(z_p) * a * k_p ^ (a - 1) * inv(c_p) * inv(c_p)
```

**Comment by wupeifan** [comment]
I don't think it's a missing method problem, or, at least, the probability is low... 

Let's not put the comments on the potential bugs to MTK and the implementation difficulties of our project together here. I think only the previous one matters here, right?
The purpose of the new interface that we are writing is exactly to treat some parameters as variables. Apparently MTK treats parameters and variables differently...

**Comment by wupeifan** [comment]
> For example, here it is in the REPL, which is fine... but when the expression is the output of a ModelingToolkit.jacobian() call inside of a loop... then things break.

Sorry I missed this one. Is this the MWE to generate the error?

**Comment by wupeifan** [comment]
@arnavs Is your issue similar to what I have here?

```
using ModelingToolkit
@variables c c_p z z_p k k_p
@parameters r b a sigma_2
F_stack = Vector{Operation}(undef, 2)
F_stack[1] = ((b * a * exp(z_p) * k_p ^ (a - 1)) / c_p) / c_p
F_stack[2] = b * a * exp(z_p)

julia> ModelingToolkit.gradient(F_stack[1], [a])
1-element Array{Operation,1}:
 inv(c_p) * (exp(z_p) * b * k_p ^ (a - 1) + b * a * exp(z_p) * k_p ^ (a - 1) * log(k_p)) * inv(c_p)

julia> ModelingToolkit.gradient(F_stack[1], [b])
1-element Array{Operation,1}:
 exp(z_p) * a * k_p ^ (a - 1) * inv(c_p) * inv(c_p)

julia> ModelingToolkit.gradient(F_stack[1], [r])
1-element Array{ModelingToolkit.Constant,1}:
 ModelingToolkit.Constant(0)

julia> ModelingToolkit.gradient(F_stack[1], [sigma_2])
1-element Array{ModelingToolkit.Constant,1}:
 ModelingToolkit.Constant(0)

# so far so good

julia> tmp = ModelingToolkit.gradient(F_stack[1], [a, r])
2-element Array{Expression,1}:
 inv(c_p) * (exp(z_p) * b * k_p ^ (a - 1) + b * a * exp(z_p) * k_p ^ (a - 1) * log(k_p)) * inv(c_p)
                                                                        ModelingToolkit.Constant(0)
# here's error comes
julia> ModelingToolkit.gradient(tmp[1], [b])
ERROR: MethodError: no method matching derivative(::typeof(*), ::Tuple{Operation,Operation,Operation}, ::Val{1})

# but the following can work
julia> ModelingToolkit.gradient(tmp[1], [r])
1-element Array{ModelingToolkit.Constant,1}:
 ModelingToolkit.Constant(0)

```

**Comment by arnavs** [comment]
@wupeifan Yeah, looks to be about right. 

The other thing to note is that the `Differential(v)(O)` object is properly constructed (@ChrisRackauckas); it's just the call to `expand_derivatives` which fails. 

This is a roadblocker for us, since we can't proceed with our work until we can take derivatives w.r.t. parameters.

**Comment by wupeifan** [comment]
@ChrisRackauckas Following the example above, it seems that "+" has a similar issue. A M(not)WE:
```
using ModelingToolkit
@variables a b c d
F = a * b * c + a * b * d + a * c * d + b * c * d
F_a = ModelingToolkit.gradient(F, [a])

julia> ModelingToolkit.gradient(F_a[1], [a, b, c, d])
ERROR: MethodError: no method matching derivative(::typeof(+), ::Tuple{Operation,Operation,Operation}, ::Val{1})
```

**Comment by wupeifan** [comment]
I found one thing (which may not be helpful or you guys already know):
Brackets are added automatically when an expression is created. For example,
```
using ModelingToolkit
@variables a b c d
julia> F = a * b * c + a * b * d + a * c * d + b * c * d
(((a * b) * c + (a * b) * d) + (a * c) * d) + (b * c) * d
```
This essentially prevents N-ary operations. Therefore, taking gradient or derivative with respect fo F directly won't trigger any error, because they are all no more than binary operations.
However, the result coming from the gradient function is not bracketed. 
```
julia> F_a = ModelingToolkit.gradient(F, [a])
1-element Array{Operation,1}:
 c * b + d * b + d * c
```
If we can apply the same construction used in `F` in `F_a`, then we can solve that problem in a cheap way.

**Comment by ChrisRackauckas** [comment]
I really think it's in the chain rule. @YingboMa I wonder if you can take a quick look and see if you can find it.

**Comment by YingboMa** [comment]
Duplicate of https://github.com/JuliaDiffEq/ModelingToolkit.jl/issues/166
